[Release Workflow] Test suite checkout fails after validate job force-pushes tag

## Problem

The release workflow (`release.yml`) fails on every v0.0.61 run. The **validate** job succeeds (creates version bump commit, force-pushes tag to version bump SHA), but the **test** job immediately fails at checkout.

## Current Behavior

1. Tag `v0.0.61` is pushed → workflow triggers
2. **Validate job** succeeds:
   - Creates `chore(release): v0.0.61 [skip ci]` commit on main
   - Creates `chore(release): restore :edge image tags [skip ci]` commit on main
   - Force-pushes tag to point to the version bump commit (per #2525)
3. **Test job** fails at checkout step:
   ```
   ##[error]The ref 'refs/tags/v0.0.61' does not point to the expected commit
   'd5d739e06cb06dc949fec04efc7228b13eda1d7b'. The ref may have been updated
   after the workflow was triggered.
   ```
4. All downstream jobs (publish-npm, publish-containers, create-release) are **skipped**

## Expected Behavior

The test suite should check out the correct ref (either the original tag commit or the version bump commit) and run tests successfully, allowing the publish pipeline to complete.

## Root Cause Analysis

The `actions/checkout@v6` action has a built-in safety check: it compares the current tag ref against `github.sha` (the SHA that triggered the workflow). Since the validate job force-pushes the tag to a **new** commit, the checkout action detects the mismatch and refuses to proceed.

This is a race condition by design:
- `github.sha` = original commit when workflow was triggered (`d5d739e0`)
- Tag now points to = version bump commit (`0dea8716`)
- Checkout action rejects because `d5d739e0 != 0dea8716`

## Evidence

Two consecutive runs both failed identically:
- Run 23100433011 (01:15:30 UTC) — failure
- Run 23100411068 (01:14:05 UTC) — failure

Previous releases (v0.0.60, v0.0.59, v0.0.58) all succeeded, suggesting this may be a regression in `actions/checkout@v6` behavior or a subtle ordering change.

## Steps to Reproduce

1. Push any new semver tag: `git tag v0.0.62 && git push origin v0.0.62`
2. Observe validate job succeeds
3. Observe test job fails at checkout with ref mismatch error

## Proposed Fix

The test job (line 421-422 of `release.yml`) uses a bare checkout without specifying `ref`:
```yaml
- name: Checkout repository
  uses: actions/checkout@v6
```

This defaults to `github.ref` (the tag), which has been force-pushed by the validate job. Options:

1. **Pin checkout to `github.sha`**: Add `ref: ${{ github.sha }}` to the test job checkout — this checks out the original commit that was tagged, ignoring the tag repoint
2. **Pin checkout to version bump SHA**: Add `ref: ${{ needs.validate.outputs.version_bump_sha }}` — checks out the version bump commit directly
3. **Check out main branch**: Add `ref: main` — since version bump is on main, this gets the latest state

Option 1 is simplest and safest — it preserves the original intent (test the code that was tagged).

## Context

- Branch: main
- Failed tag: v0.0.61
- Original commit: d5d739e06cb06dc949fec04efc7228b13eda1d7b
- Version bump commit: 0dea8716
- Edge restore commit: 040fe7d8
- Relevant files: `.github/workflows/release.yml` (lines 421-422)
- Related issues: #2525, #2526, #2534, #2576, #2577, #2578

--- a/.github/workflows/release.yml
+++ b/.github/workflows/release.yml
@@ -418,8 +418,14 @@ jobs:
       contents: read
 
     steps:
+      # #2588 — Use github.sha (the commit that was tagged when the workflow
+      # was triggered) instead of the default ref. The validate job force-pushes
+      # the tag to the version bump commit (#2525), which causes checkout@v6 to
+      # reject the default ref with "ref may have been updated after trigger".
       - name: Checkout repository
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
+        with:
+          ref: ${{ github.sha }}
 
       - name: Build + start devcontainer services
         run: |
@@ -516,8 +522,12 @@ jobs:
       id-token: write  # npm provenance
 
     steps:
+      # #2588 — Pin to github.sha to avoid checkout@v6 rejection after
+      # validate force-pushes the tag. Artifact overlay provides version-bumped files.
       - name: Checkout repository
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
+        with:
+          ref: ${{ github.sha }}
 
       - name: Download bumped version files
         uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c # v8.0.1
@@ -602,8 +612,12 @@ jobs:
       packages: write
 
     steps:
+      # #2588 — Pin to github.sha to avoid checkout@v6 rejection after
+      # validate force-pushes the tag. Artifact overlay provides version-bumped files.
       - name: Checkout repository
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
+        with:
+          ref: ${{ github.sha }}
 
       - name: Download bumped version files
         uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c # v8.0.1
@@ -714,8 +728,12 @@ jobs:
             dockerfile: docker/symphony-worker/Dockerfile
 
     steps:
+      # #2588 — Pin to github.sha to avoid checkout@v6 rejection after
+      # validate force-pushes the tag. Artifact overlay provides version-bumped files.
       - name: Checkout repository
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
+        with:
+          ref: ${{ github.sha }}
 
       - name: Download bumped version files
         uses: actions/download-artifact@3e5f45b2cfb9172054b4087a40e8e0b5a5461e7c # v8.0.1
@@ -839,10 +857,13 @@ jobs:
             echo "::notice::Sentry not configured — skipping release finalization"
           fi
 
+      # #2588 — Pin to github.sha to avoid checkout@v6 rejection after
+      # validate force-pushes the tag.
       - name: Checkout repository
         if: steps.sentry-check.outputs.configured == 'true'
         uses: actions/checkout@de0fac2e4500dabe0009e67214ff5f5447ce83dd # v6
         with:
+          ref: ${{ github.sha }}
           fetch-depth: 0
 
       - name: Setup Node.js

